plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,759 @@
+"""
+Module containing the definition for the general particle tracker.
+"""
+
+__all__ = [
+    "AbstractDiskSaveRoutine",
+    "AbstractIntervalSaveRoutine",
+    "AbstractMemorySaveRoutine",
+    "AbstractSaveRoutine",
+    "AbstractStopCondition",
+    "DiskIntervalSaveRoutine",
+    "MemoryIntervalSaveRoutine",
+    "ParticleTracker",
+    "TimeElapsedStopCondition",
+]
+
+import astropy.units as u
+import collections
+import h5py
+import numpy as np
+import sys
+import warnings
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from pathlib import Path
+from tqdm import tqdm
+from typing import Optional, Union
+
+from plasmapy.particles import Particle, particle_input
+from plasmapy.plasma.grids import AbstractGrid
+from plasmapy.plasma.plasma_base import BasePlasma
+from plasmapy.simulation.particle_integrators import boris_push
+
+
+class AbstractStopCondition(ABC):
+    """Abstract base class containing the necessary methods for a ParticleTracker stopping condition."""
+
+    @abstractmethod
+    def require_uniform_dt(self):
+        """Return whether or not this stop condition requires a uniform dt to be specified."""
+        ...
+
+    @abstractmethod
+    def get_description(self):
+        """Return a small string describing the relevant quantity shown on the meter."""
+        ...
+
+    @abstractmethod
+    def get_units(self):
+        """Return the units of `total`."""
+        ...
+
+    @abstractmethod
+    def is_finished(self, particle_tracker):
+        """Return `True` if the simulation has finished."""
+        ...
+
+    @abstractmethod
+    def get_progress(self, particle_tracker):
+        """Return a number representing the progress of the simulation (compared to total).
+        This number represents the numerator of the completion percentage.
+        """
+        ...
+
+    @abstractmethod
+    def get_total(self, particle_tracker):
+        """Return a number representing the total number of steps in a simulation.
+        This number represents the denominator of the completion percentage.
+        """
+        ...
+
+
+class TimeElapsedStopCondition(AbstractStopCondition):
+    """Stop condition corresponding to the elapsed time of a ParticleTracker."""
+
+    def __init__(self, stop_time: u.Quantity):
+        self.stop_time = stop_time.to(u.s).value
+
+    @property
+    def require_uniform_dt(self):
+        """The elapsed time stop condition requires a uniform time step
+        to stop all particles at the same time.
+        """
+        return True
+
+    @property
+    def get_description(self):
+        """The time elapsed stop condition depends on elapsed time,
+        therefore the relevant quantity is time remaining.
+        """
+        return "Time remaining"
+
+    @property
+    def get_units(self):
```

Comment:
I would return an astropy object here, `u.s`

Reply:
I get `TypeError: Unsupported operand type(s) for ufunc add: 'str,Quantity'`